README.md
# hexo-adsense
Hexo adsense support

> first development finished on 12 November 2021

# Features
- can display adsense to all pages except specific posts or pages (multiple supported)
- can only display adsense to specific posts or pages
- adsense javascript enchantments (lazy adsense included)
- support display adsense on localhost

# Installation
Using Git Repository (Development)
```shell
npm i git+https://github.com/dimaslanjaka/hexo-adsense.git
```
Using NPM Repository (Production)
```shell
npm i hexo-adsense
```

### _config.yml
```yaml
adsense:
  # enable(true) or disable(false)
  enable: true
  # ca-pub-xxxx
  pub: ca-pub-1165447249910969
  # in-article ads source
  article_ads:
    - "source/ads/in_article.html"
    - "source/ads/in_article2.html"
  # ads field
  field: "site" # post=only post, site=all pages and posts
  # auto redirect to https
  https: true
  # block user with adblock enabled
  adblock: true
  # exclude pattern to disable adsense
  exclude:
    - "*.min.html"
    - "exclude/**/*"
```

### Article Ads
> include to config article_ads
```yaml
adsense:
  article_ads:
    - "source/ads/in_article.html"
```
> `source/ads/in_article.html` contents
```html
<!--REMOVE ADSENSE SCRIPT, THIS PLUGIN ALREADY OPTIMIZED THE ADSENSE JAVASCRIPT-->
<!--<script async src="//pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-xxxx" crossorigin="anonymous"></script>-->
<ins class="adsbygoogle"
     style="display:block; text-align:center;"
     data-ad-layout="in-article"
     data-ad-format="fluid"
     data-ad-client="ca-pub-xxxxx"
     data-ad-slot="xxxxx"></ins>
<!--script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script-->
```

# disable options
> to disable adsense on pattern `_config.yml` or set each post `adsense: false`.

<hr>

### Enable all except specific post
> enable adsense on all pages except spescific posts or pages.
> set `_config.yml`
```yaml
adsense:
  enable: true
  field: "site"
  #... pub, article_ads, etc
```
> set post header
```yaml
title: "Post title"
adsense: false
tags:
  - tags1
  - tags2
```

<hr>

### Disable all except post
> disable adsense on all pages and posts, except specific posts
> set `_config.yml`
```yaml
adsense:
  enable: false # adsense disabled globally
  #... pub, article_ads, etc
```
> set post header
```yaml
title: "Post title"
adsense: true # adsense will shown only for post with header `adsense: true` / enabled
tags:
  - tags1
  - tags2
```

# Website using Hexo NodeJS Blogging System

[![Build And Tests](https://github.com/dimaslanjaka/dimaslanjaka.github.io/actions/workflows/page.yml/badge.svg?branch=compiler)](https://github.com/dimaslanjaka/dimaslanjaka.github.io/actions/workflows/page.yml)
[![GitHub](https://badgen.net/badge/icon/github?icon=github&label&style=flat-square)](https://github.com/dimaslanjaka/dimaslanjaka.github.io/tree/compiler)
[![webmanajemen.com](https://img.shields.io/website.svg?down_color=red&down_message=down&style=flat-square&up_color=green&up_message=up&label=webmanajemen.com&url=https://webmanajemen.com)](https://webmanajemen.com)

## hexo-adsense
[![npm version](https://badge.fury.io/js/hexo-adsense.svg?style=flat-square)](https://badge.fury.io/js/hexo-adsense)
[![Npm package yearly downloads](https://badgen.net/npm/dy/hexo-adsense?style=flat-square)](https://npmjs.com/package/hexo-adsense)
[![Minimum node.js version](https://badgen.net/npm/node/hexo-adsense?style=flat-square)](https://npmjs.com/package/hexo-adsense)
![GitHub repo size](https://img.shields.io/github/repo-size/dimaslanjaka/hexo-adsense?label=Repository%20Size&style=flat-square)
![GitHub last commit](https://img.shields.io/github/last-commit/dimaslanjaka/hexo-adsense?color=blue&label=Last%20Commit&style=flat-square)

## hexo-seo
[![npm version](https://badge.fury.io/js/hexo-seo.svg?style=flat-square)](https://badge.fury.io/js/hexo-seo)
[![Npm package yearly downloads](https://badgen.net/npm/dy/hexo-seo?style=flat-square)](https://npmjs.com/package/hexo-seo)
[![Minimum node.js version](https://badgen.net/npm/node/hexo-seo?style=flat-square)](https://npmjs.com/package/hexo-seo)
![GitHub repo size](https://img.shields.io/github/repo-size/dimaslanjaka/hexo-seo?label=Repository%20Size&style=flat-square)
![GitHub last commit](https://img.shields.io/github/last-commit/dimaslanjaka/hexo-seo?color=blue&label=Last%20Commit&style=flat-square)

## hexo-blogger-xml
[![npm version](https://badge.fury.io/js/hexo-blogger-xml.svg?style=flat-square)](https://badge.fury.io/js/hexo-blogger-xml)
[![Npm package yearly downloads](https://badgen.net/npm/dy/hexo-blogger-xml?style=flat-square)](https://npmjs.com/package/hexo-blogger-xml)
[![Minimum node.js version](https://badgen.net/npm/node/hexo-blogger-xml?style=flat-square)](https://npmjs.com/package/hexo-blogger-xml)
![GitHub repo size](https://img.shields.io/github/repo-size/dimaslanjaka/hexo-blogger-xml?label=Repository%20Size&style=flat-square)
![GitHub last commit](https://img.shields.io/github/last-commit/dimaslanjaka/hexo-blogger-xml?color=blue&label=Last%20Commit&style=flat-square)


CHANGELOG.md
# CHANGELOG

## [0.0.3] November 11, 2021
## [0.0.4] November 11, 2021
## [0.0.5] November 11, 2021
- f067bee : -debug
- d27622a : +cache strategy

## [0.0.6] November 11, 2021
- e621b9b : fix publish changelog
- b0ed3b7 : fix publish changelog
- df83126 : Test changelog

## [0.0.7] November 11, 2021
- 226e5a8 : +information package

## [0.0.8] November 11, 2021
- 2dc9bc6 : remove submodule
- 9de7933 : Update release 0.0.8
- c04e3f0 : change source
- 71646c9 : remove submodule

## [0.0.9] November 11, 2021
- edeb9aa : update docs
- d53f764 : Update docs
- a523b6a : Update docs
- 9de7933 : Update release 0.0.8
- c04e3f0 : change source
- 71646c9 : remove submodule

## [0.0.10] November 11, 2021
- d15ac4e : fix memory cache stategy
- 9de7933 : Update release 0.0.8
- c04e3f0 : change source
- 71646c9 : remove submodule

## [0.0.11] November 11, 2021
- 401d1d6 : remove prepare script
- 9de7933 : Update release 0.0.8
- c04e3f0 : change source
- 71646c9 : remove submodule

## [0.0.12] November 11, 2021
- 61083bd : fix field post
- ad63157 : reduce filter_adshtml arguments
- 781737a : fix hexo unparsed
- 66913e9 : +advance filtering and cache strategy
- 9de7933 : Update release 0.0.8
- c04e3f0 : change source
- 71646c9 : remove submodule

## [0.0.13] November 12, 2021
- 1100087 : fix adsense injected on non-html files
- 9de7933 : Update release 0.0.8
- c04e3f0 : change source
- 71646c9 : remove submodule

## [0.0.14] November 12, 2021
- 2288565 : +cache
- 26ddbb2 : fix adsense not display on archives etc
- 9de7933 : Update release 0.0.8
- c04e3f0 : change source
- 71646c9 : remove submodule

## [0.0.15] November 13, 2021
- 12fc85f : fix adsense not showing on posts
- 9de7933 : Update release 0.0.8
- c04e3f0 : change source
- 71646c9 : remove submodule

## [0.0.16] November 13, 2021
- 2102d6e : add pre selector on header process source/article-ads.js:113
- Sat Nov 13 17:02:53 2021 +0700
- 1e3c150 : fix adsense displayed on em,strong  Sat Nov 13 16:56:02 2021 +0700
- 63edb45 : fix adsense displayed inside <pre>/<code>  Sat Nov 13 16:44:29 2021 +0700
- 9de7933 : Update release 0.0.8  Thu Nov 11 19:42:53 2021 +0700
- c04e3f0 : change source  Thu Nov 11 19:42:39 2021 +0700
- 71646c9 : remove submodule  Thu Nov 11 19:42:15 2021 +0700

## [0.0.17] November 15, 2021
- baa50c6 : cancel adsense if config not defined  Mon Nov 15 13:29:53 2021 +0700

## [0.0.19] 2021-11-15T17:29:01
- 98dc5fe : migrate using typescript compiler  Mon Nov 15 17:28:33 2021 +0700

## [0.0.20] 2021-11-15T20:19:41
- 6ecbcf3 : fix source files ENOENT  Mon Nov 15 20:19:05 2021 +0700
- 2d12d3f : fix source files ENOENT  Mon Nov 15 20:18:56 2021 +0700
- 6e6b4af : fix source css/js not found  Mon Nov 15 20:17:32 2021 +0700
- 9de7933 : Update release 0.0.8  Thu Nov 11 19:42:53 2021 +0700
- c04e3f0 : change source  Thu Nov 11 19:42:39 2021 +0700
- 71646c9 : remove submodule  Thu Nov 11 19:42:15 2021 +0700

## [1.0.21] 2021-11-15T22:57:36
- 731625d : update environment  Mon Nov 15 22:56:18 2021 +0700

## [1.0.22] 2021-11-15T23:00:52
- 9ad891e : update browser config  Mon Nov 15 23:00:25 2021 +0700

## [1.0.23] 2021-11-15T23:51:55
- c5fcbc5 : update packaging  Mon Nov 15 23:51:33 2021 +0700
- 9de7933 : Update release 0.0.8  Thu Nov 11 19:42:53 2021 +0700
- c04e3f0 : change source  Thu Nov 11 19:42:39 2021 +0700
- 71646c9 : remove submodule  Thu Nov 11 19:42:15 2021 +0700

## [1.0.24] 2021-11-16T09:08:44
- 62c0568 : reduce inMemory cache usage  Tue Nov 16 09:08:20 2021 +0700
- 4116d6e : reduce in-memory cache usage  Tue Nov 16 09:07:31 2021 +0700
- 9de7933 : Update release 0.0.8  Thu Nov 11 19:42:53 2021 +0700
- c04e3f0 : change source  Thu Nov 11 19:42:39 2021 +0700
- 71646c9 : remove submodule  Thu Nov 11 19:42:15 2021 +0700

## [1.0.25] 2021-11-26T23:51:28
- 8d42898 : Fix minifier  Fri Nov 26 23:47:04 2021 +0700
- 11050f9 : reduce resources  Fri Nov 26 23:13:04 2021 +0700
- a4ab023 : minify  Fri Nov 26 23:06:26 2021 +0700
- 3a7f3ec : minify source  Fri Nov 26 22:23:39 2021 +0700
- 9de7933 : Update release 0.0.8  Thu Nov 11 19:42:53 2021 +0700
- c04e3f0 : change source  Thu Nov 11 19:42:39 2021 +0700
- 71646c9 : remove submodule  Thu Nov 11 19:42:15 2021 +0700

## [1.0.26] 2021-11-27T10:27:27
- a2e2b44 : Fix update  Sat Nov 27 10:26:47 2021 +0700
- 9de7933 : Update release 0.0.8  Thu Nov 11 19:42:53 2021 +0700
- c04e3f0 : change source  Thu Nov 11 19:42:39 2021 +0700
- 71646c9 : remove submodule  Thu Nov 11 19:42:15 2021 +0700


src/md5.js
function md5cycle(x, k) {
  var a = x[0],
    b = x[1],
    c = x[2],
    d = x[3];

  a = ff(a, b, c, d, k[0], 7, -680876936);
  d = ff(d, a, b, c, k[1], 12, -389564586);
  c = ff(c, d, a, b, k[2], 17, 606105819);
  b = ff(b, c, d, a, k[3], 22, -1044525330);
  a = ff(a, b, c, d, k[4], 7, -176418897);
  d = ff(d, a, b, c, k[5], 12, 1200080426);
  c = ff(c, d, a, b, k[6], 17, -1473231341);
  b = ff(b, c, d, a, k[7], 22, -45705983);
  a = ff(a, b, c, d, k[8], 7, 1770035416);
  d = ff(d, a, b, c, k[9], 12, -1958414417);
  c = ff(c, d, a, b, k[10], 17, -42063);
  b = ff(b, c, d, a, k[11], 22, -1990404162);
  a = ff(a, b, c, d, k[12], 7, 1804603682);
  d = ff(d, a, b, c, k[13], 12, -40341101);
  c = ff(c, d, a, b, k[14], 17, -1502002290);
  b = ff(b, c, d, a, k[15], 22, 1236535329);

  a = gg(a, b, c, d, k[1], 5, -165796510);
  d = gg(d, a, b, c, k[6], 9, -1069501632);
  c = gg(c, d, a, b, k[11], 14, 643717713);
  b = gg(b, c, d, a, k[0], 20, -373897302);
  a = gg(a, b, c, d, k[5], 5, -701558691);
  d = gg(d, a, b, c, k[10], 9, 38016083);
  c = gg(c, d, a, b, k[15], 14, -660478335);
  b = gg(b, c, d, a, k[4], 20, -405537848);
  a = gg(a, b, c, d, k[9], 5, 568446438);
  d = gg(d, a, b, c, k[14], 9, -1019803690);
  c = gg(c, d, a, b, k[3], 14, -187363961);
  b = gg(b, c, d, a, k[8], 20, 1163531501);
  a = gg(a, b, c, d, k[13], 5, -1444681467);
  d = gg(d, a, b, c, k[2], 9, -51403784);
  c = gg(c, d, a, b, k[7], 14, 1735328473);
  b = gg(b, c, d, a, k[12], 20, -1926607734);

  a = hh(a, b, c, d, k[5], 4, -378558);
  d = hh(d, a, b, c, k[8], 11, -2022574463);
  c = hh(c, d, a, b, k[11], 16, 1839030562);
  b = hh(b, c, d, a, k[14], 23, -35309556);
  a = hh(a, b, c, d, k[1], 4, -1530992060);
  d = hh(d, a, b, c, k[4], 11, 1272893353);
  c = hh(c, d, a, b, k[7], 16, -155497632);
  b = hh(b, c, d, a, k[10], 23, -1094730640);
  a = hh(a, b, c, d, k[13], 4, 681279174);
  d = hh(d, a, b, c, k[0], 11, -358537222);
  c = hh(c, d, a, b, k[3], 16, -722521979);
  b = hh(b, c, d, a, k[6], 23, 76029189);
  a = hh(a, b, c, d, k[9], 4, -640364487);
  d = hh(d, a, b, c, k[12], 11, -421815835);
  c = hh(c, d, a, b, k[15], 16, 530742520);
  b = hh(b, c, d, a, k[2], 23, -995338651);

  a = ii(a, b, c, d, k[0], 6, -198630844);
  d = ii(d, a, b, c, k[7], 10, 1126891415);
  c = ii(c, d, a, b, k[14], 15, -1416354905);
  b = ii(b, c, d, a, k[5], 21, -57434055);
  a = ii(a, b, c, d, k[12], 6, 1700485571);
  d = ii(d, a, b, c, k[3], 10, -1894986606);
  c = ii(c, d, a, b, k[10], 15, -1051523);
  b = ii(b, c, d, a, k[1], 21, -2054922799);
  a = ii(a, b, c, d, k[8], 6, 1873313359);
  d = ii(d, a, b, c, k[15], 10, -30611744);
  c = ii(c, d, a, b, k[6], 15, -1560198380);
  b = ii(b, c, d, a, k[13], 21, 1309151649);
  a = ii(a, b, c, d, k[4], 6, -145523070);
  d = ii(d, a, b, c, k[11], 10, -1120210379);
  c = ii(c, d, a, b, k[2], 15, 718787259);
  b = ii(b, c, d, a, k[9], 21, -343485551);

  x[0] = add32(a, x[0]);
  x[1] = add32(b, x[1]);
  x[2] = add32(c, x[2]);
  x[3] = add32(d, x[3]);
}

function cmn(q, a, b, x, s, t) {
  a = add32(add32(a, q), add32(x, t));
  return add32((a << s) | (a >>> (32 - s)), b);
}

function ff(a, b, c, d, x, s, t) {
  return cmn((b & c) | (~b & d), a, b, x, s, t);
}

function gg(a, b, c, d, x, s, t) {
  return cmn((b & d) | (c & ~d), a, b, x, s, t);
}

function hh(a, b, c, d, x, s, t) {
  return cmn(b ^ c ^ d, a, b, x, s, t);
}

function ii(a, b, c, d, x, s, t) {
  return cmn(c ^ (b | ~d), a, b, x, s, t);
}

function md51(s) {
  txt = "";
  var n = s.length,
    state = [1732584193, -271733879, -1732584194, 271733878],
    i;
  for (i = 64; i <= s.length; i += 64) {
    md5cycle(state, md5blk(s.substring(i - 64, i)));
  }
  s = s.substring(i - 64);
  var tail = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
  for (i = 0; i < s.length; i++) tail[i >> 2] |= s.charCodeAt(i) << (i % 4 << 3);
  tail[i >> 2] |= 0x80 << (i % 4 << 3);
  if (i > 55) {
    md5cycle(state, tail);
    for (i = 0; i < 16; i++) tail[i] = 0;
  }
  tail[14] = n * 8;
  md5cycle(state, tail);
  return state;
}

/* there needs to be support for Unicode here,
 * unless we pretend that we can redefine the MD-5
 * algorithm for multi-byte characters (perhaps
 * by adding every four 16-bit characters and
 * shortening the sum to 32 bits). Otherwise
 * I suggest performing MD-5 as if every character
 * was two bytes--e.g., 0040 0025 = @%--but then
 * how will an ordinary MD-5 sum be matched?
 * There is no way to standardize text to something
 * like UTF-8 before transformation; speed cost is
 * utterly prohibitive. The JavaScript standard
 * itself needs to look at this: it should start
 * providing access to strings as preformed UTF-8
 * 8-bit unsigned value arrays.
 */
function md5blk(s) {
  /* I figured global was faster.   */
  var md5blks = [],
    i; /* Andy King said do it this way. */
  for (i = 0; i < 64; i += 4) {
    md5blks[i >> 2] =
      s.charCodeAt(i) + (s.charCodeAt(i + 1) << 8) + (s.charCodeAt(i + 2) << 16) + (s.charCodeAt(i + 3) << 24);
  }
  return md5blks;
}

var hex_chr = "0123456789abcdef".split("");

function rhex(n) {
  var s = "",
    j = 0;
  for (; j < 4; j++) s += hex_chr[(n >> (j * 8 + 4)) & 0x0f] + hex_chr[(n >> (j * 8)) & 0x0f];
  return s;
}

function hex(x) {
  for (var i = 0; i < x.length; i++) x[i] = rhex(x[i]);
  return x.join("");
}

function md5(s) {
  if (typeof s == "undefined") s = "undefined";
  if (!s) s = "null";
  return hex(md51(s));
}

/* this function is much faster,
  so if possible we use it. Some IEs
  are the only ones I know of that
  need the idiotic second function,
  generated by an if clause.  */

function add32(a, b) {
  return (a + b) & 0xffffffff;
}

if (md5("hello") != "5d41402abc4b2a76b9719d911017c592") {
  function add32(x, y) {
    var lsw = (x & 0xffff) + (y & 0xffff),
      msw = (x >> 16) + (y >> 16) + (lsw >> 16);
    return (msw << 16) | (lsw & 0xffff);
  }
}

module.exports = md5;
